src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,171 @@
+        {
+            TypeSyntax declaredType;
+            State state = null;
+            var shouldAnalyze = false;
+            var declarationStatement = context.Node;
+            var optionSet = GetOptionSet(context.Options);
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+
+            if (declarationStatement.IsKind(SyntaxKind.VariableDeclaration))
+            {
+                var declaration = (VariableDeclarationSyntax)declarationStatement;
+                declaredType = declaration.Type;
+
+                shouldAnalyze = ShouldAnalyzeVariableDeclaration(declaration, semanticModel, cancellationToken);
+
+                if (shouldAnalyze)
+                {
+                    state = State.Generate(declarationStatement, semanticModel, optionSet, isVariableDeclarationContext: true, cancellationToken: cancellationToken);
+                    shouldAnalyze = IsStylePreferred(semanticModel, optionSet, state, cancellationToken);
+                }
+            }
+            else if (declarationStatement.IsKind(SyntaxKind.ForEachStatement))
+            {
+                var declaration = (ForEachStatementSyntax)declarationStatement;
+                declaredType = declaration.Type;
+
+                state = State.Generate(declarationStatement, semanticModel, optionSet, isVariableDeclarationContext: false, cancellationToken: cancellationToken);
+                shouldAnalyze = IsStylePreferred(semanticModel, optionSet, state, cancellationToken);
+            }
+            else
+            {
+                Debug.Assert(false, $"called in for unregistered node kind {declarationStatement.Kind().ToString()}");
+                return;
+            }
+
+            if (shouldAnalyze)
+            {
+                Debug.Assert(state != null, "analyzing a declaration and state is null.");
+
+                TextSpan diagnosticSpan;
+
+                if (TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken, out diagnosticSpan))
+                {
+                    var descriptor = _severityToDescriptorMap[state.GetDiagnosticSeverityPreference()];
+                    context.ReportDiagnostic(CreateDiagnostic(descriptor, declarationStatement, diagnosticSpan));
+                }
+            }
+        }
+
+        private Diagnostic CreateDiagnostic(DiagnosticDescriptor descriptor, SyntaxNode declaration, TextSpan diagnosticSpan) =>
+            Diagnostic.Create(descriptor, declaration.SyntaxTree.GetLocation(diagnosticSpan));
+
+        private bool ShouldAnalyzeVariableDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // var is applicable only for local variables.
+            if (variableDeclaration.IsParentKind(SyntaxKind.FieldDeclaration,
```

Comment:
If the compiler adds another location, I or whoever supports ide features has to touch this lost -- to specifically include or exclude that kind. But yeah, with a whitelist missing a kind, it's restricted to feature not working there and we get away without a crash. I'll change this in a bit and take your if branch that yields the :+1 ;)

Sent from Outlook Mobile<https://aka.ms/blhgte>

From: CyrusNajmabadi
Sent: Wednesday, February 24, 5:06 PM
Subject: Re: [roslyn] Feature: Suggest implicit or explicit typing code styles. (#8045)
To: dotnet/roslyn
Cc: Balaji Krishnan

In src<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>/Features/<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>CSharp<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>/Portable/Diagnostics/Analyzers/<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>CSharpTypingStyleDiagnosticAnalyzerBase.cs<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>:

> + > + if (TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken, out diagnosticSpan)) > + { > + var descriptor = _severityToDescriptorMap[state.GetDiagnosticSeverityPreference()]; > + context.ReportDiagnostic(CreateDiagnostic(descriptor, declarationStatement, diagnosticSpan)); > + } > + } > + } > + > + private Diagnostic CreateDiagnostic(DiagnosticDescriptor descriptor, SyntaxNode declaration, TextSpan diagnosticSpan) => > + Diagnostic.Create(descriptor, declaration.SyntaxTree.GetLocation(diagnosticSpan)); > + > + private bool ShouldAnalyzeVariableDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken) > + { > + // var is applicable only for local variables. > + if (variableDeclaration.IsParentKind(SyntaxKind.FieldDeclaration,

This needs to be a whitelist. How do you know that just filtering out fields will be enough? What happens when the compiler adds another location where VairableDeclarations exist?

-
Reply to this email directly or view it on <https://github.com/dotnet/roslyn/pull/8045/files#r54034687> GitHub<https://github.com/dotnet/roslyn/pull/8045/files#r54034687>.